File: leetcode/Medium/0641-design-circular-deque/0641-design-circular-deque.java
class MyCircularDeque {
  private List<Integer> q;
  private int x;

  public MyCircularDeque(int k) {
    q = new LinkedList<Integer>();
    x = k;
  }

  public boolean insertFront(int value) {
    if (q.size() == x) {
      return false;
    } else {
      q.addFirst(value);
      return true;
    }
  }

  public boolean insertLast(int value) {
    if (q.size() == x) {
      return false;
    } else {
      q.addLast(value);
      return true;
    }
  }

  public boolean deleteFront() {
    if (q.size() == 0) {
      return false;
    } else {
      q.removeFirst();
      return true;
    }
  }

  public boolean deleteLast() {
    if (q.size() == 0) {
      return false;
    } else {
      q.removeLast();
      return true;
    }
  }

  public int getFront() {
    if (q.isEmpty()) {
      return -1;
    }
    return q.getFirst();
  }

  public int getRear() {
    if (q.isEmpty()) {
      return -1;
    }
    return q.getLast();
  }

  public boolean isEmpty() {
    return q.isEmpty();
  }

  public boolean isFull() {
    if (q.size() == x) {
      return true;
    } else {
      return false;
    }
  }
}

/**
 * Your MyCircularDeque object will be instantiated and called as such: MyCircularDeque obj = new
 * MyCircularDeque(k); boolean param_1 = obj.insertFront(value); boolean param_2 =
 * obj.insertLast(value); boolean param_3 = obj.deleteFront(); boolean param_4 = obj.deleteLast();
 * int param_5 = obj.getFront(); int param_6 = obj.getRear(); boolean param_7 = obj.isEmpty();
 * boolean param_8 = obj.isFull();
 */

File: leetcode/Medium/0641-design-circular-deque/0641-design-circular-deque.cpp
class MyCircularDeque {
public:
    list<int>q;
    int x;
    /** Initialize your data structure here. Set the size of the deque to be k. */
    MyCircularDeque(int k) {
        list<int>q(k);
        x=k;
    }
    
    /** Adds an item at the front of Deque. Return true if the operation is successful. */
    bool insertFront(int value) {
        if(q.size()==x)
            return false;
        else{
            q.insert(q.begin(), value);
            return true;
        }
           
    }
    
    /** Adds an item at the rear of Deque. Return true if the operation is successful. */
    bool insertLast(int value) {
        if(q.size()==x)
            return false;
        else{
            q.insert(q.end(), value);
            return true;
        }
    }
    
    /** Deletes an item from the front of Deque. Return true if the operation is successful. */
    bool deleteFront() {
        if(q.size()==0)
            return false;
        else{
            q.pop_front();
            return true;
        }
    }
    
    /** Deletes an item from the rear of Deque. Return true if the operation is successful. */
    bool deleteLast() {
        if(q.size()==0)
            return false;
        else{
            q.pop_back();
            return true;
        }
    }
    
    /** Get the front item from the deque. */
    int getFront() {
        if(q.size()==0)
            return -1;
        else
            return q.front();
    }
    
    /** Get the last item from the deque. */
    int getRear() {
        if(q.size()==0)
            return -1;
        else
            return q.back();
    }
    
    /** Checks whether the circular deque is empty or not. */
    bool isEmpty() {
        return q.size()==0;
    }
    
    /** Checks whether the circular deque is full or not. */
    bool isFull() {
        return q.size()==x;
    }
};

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * MyCircularDeque* obj = new MyCircularDeque(k);
 * bool param_1 = obj->insertFront(value);
 * bool param_2 = obj->insertLast(value);
 * bool param_3 = obj->deleteFront();
 * bool param_4 = obj->deleteLast();
 * int param_5 = obj->getFront();
 * int param_6 = obj->getRear();
 * bool param_7 = obj->isEmpty();
 * bool param_8 = obj->isFull();
 */